seq2simulate/run_make_proviral_data.py

```diff
@@ -0,0 +1,447 @@
+            if cod == i:
+                print('{}: {}'.format(cod, mut))
+
+    return True
+
+def _make_mutation_data_files(sequences, working_dir, hypermutation_rate=3):
+
+    """
+    Performs mutations of various kinds on template sequence(s) and saves both
+    the mutated sequences and a JSON of the mutation locations to file. Returns
+    these files in two dictionaries.
+    """
+    data_files = {}
+    diff_files = {}
+
+    mutation_types  = ['hypermutation',
+                       'longdel', 
+                       'insertion', 
+                       'frameshift', 
+                       'stopcodon', 
+                       'inversion']
+
+    mutation_funcs = {'hypermutation': [diversity._simulate_hypermutation, {'hypermutation_rate': hypermutation_rate}],
+                      'longdel': [diversity._simulate_deletions, {'freq': 1, 'no_frameshifts': True}], 
+                      'insertion': [diversity._simulate_insertions, {'freq': 1, 'no_frameshifts': True}], 
+                      'frameshift': [diversity._simulate_frameshifts, {'freq': 1}],
+                      'stopcodon': [diversity._simulate_stop_codons, {'freq': 1}],
+                      'inversion': [diversity._simulate_inversions, {'freq': 1}],
+                     }
+
+    for i_mutation, mutation_type in enumerate(mutation_types):
+        mutation_func, mutation_kwargs = mutation_funcs[mutation_type]
+        sequences_strings, sequence_ids = diversity._convert_seqs_to_strs(sequences)
+        sequences_strings, seq_diffs = mutation_func(sequences_strings, **mutation_kwargs)
+        seq_diffs = dict(zip(sequence_ids, seq_diffs))
+        diversity._update_seq_reads_from_strs(sequences, sequences_strings)
+        data_files[mutation_type] = (_write_to_FASTA(sequences, working_dir, '{}_data.fasta'.format(i_mutation)))
+        diff_files[mutation_type] = (_write_to_file(seq_diffs, working_dir, '{}_diffs.data'.format(i_mutation)))
+
+    return data_files, diff_files
+
+
+def _make_art_files(data_files, platform, working_dir, paired_end=False):
+    platf = getattr(plat, platform)
+
+    fastq_files, sam_files = {}, {}
+    for mutation_type, data_file in data_files.iteritems():
+        fastq_file, sam_file = art.simulate(data_file, platf, platf.coverage, paired_end, working_dir)
+        if not isinstance(fastq_file, tuple):
+            fastq_file = [fastq_file]
+        fastq_files[mutation_type], sam_files[mutation_type] = [i for i in fastq_file], sam_file
+    return fastq_files, sam_files
+
+
+def _open_all_data_files(sequences, fastq_files, sam_files, diff_files):
+    open_fq_files = {}
+    for mutation_type, fq in fastq_files.iteritems():
+        fastq_siblings = []
+        for _fq in fq:
+            with open(_fq, 'r') as f:
+                fastq_siblings.append([i for i in parse_fastq(f)])
+        open_fq_files[mutation_type] = fastq_siblings
+    
+    open_sam_files = {}
+    for mutation_type, fs in sam_files.iteritems():
```

Comment:
You're reading a whole SAM file into memory here if I understand.  It shouldn't be too hard to phrase this as a generator, although at the file sizes we're currently doing it doesn't matter too much.  I'm torn, what do you think?